src/mongo_odm/query_builder.hpp

```diff
@@ -0,0 +1,473 @@
+     * @param  field         The value that the key is being compared to.
+     * @param  selector_type The type of comparison operator, such at gt (>) or ne (!=).
+     */
+    constexpr ComparisonExpr(const Nvp<Base, T> &nvp, T field, QuerySelector selector_type)
+        : nvp(nvp), field(std::move(field)), selector_type(selector_arr[selector_type]) {
+    }
+
+    const Nvp<Base, T> &nvp;
+    T field;
+    const std::string selector_type;
+
+    /**
+     * Appends this expression to a BSON document builder, as a key-value pair of the form
+     * "key: {$cmp: val}", where $cmp is some comparison operator.
+     * @param builder A basci BSON document builder
+     */
+    void append_to_document_builder(bsoncxx::builder::basic::sub_document &builder) const {
+        builder.append(bsoncxx::builder::basic::kvp(
+            std::string(nvp.name), [this](bsoncxx::builder::basic::sub_document subdoc) {
+                subdoc.append(bsoncxx::builder::basic::kvp(selector_type, field));
+            }));
+    }
+
+    /**
+     * Converts the expression to a BSON filter for a query.
+     * The resulting BSON is of the form "{key: {$cmp: val}}".
+     */
+    operator bsoncxx::document::view_or_value() const {
+        auto builder = bsoncxx::builder::basic::document{};
+        append_to_document_builder(builder);
+        return bsoncxx::document::view_or_value(builder);
+    }
+};
+
+/**
+ * This represents an expression with the $not operator, which wraps a comparison expression and
+ * negates it.
+ */
+template <typename Base, typename T>
+class NotExpr {
+   public:
+    /**
+     * Creates a $not expression taht negates the given comparison expression.
+     * @param  expr A comparison expression
+     */
+    constexpr NotExpr(const ComparisonExpr<Base, T> &expr) : expr(expr) {
+    }
+
+    /**
+     * Appends this expression to a BSON document builder,
+     * as a key-value pair of the form "key: {$not: {$cmp: val}}".
+     * @param builder a BSON document builder
+     */
+    void append_to_document_builder(bsoncxx::builder::basic::sub_document &builder) const {
+        builder.append(bsoncxx::builder::basic::kvp(
+            std::string(expr.nvp.name), [this](bsoncxx::builder::basic::sub_document subdoc) {
+                subdoc.append(bsoncxx::builder::basic::kvp(
+                    "$not", [this](bsoncxx::builder::basic::sub_document cmpdoc) {
+                        cmpdoc.append(bsoncxx::builder::basic::kvp(expr.selector_type, expr.field));
+                    }));
+            }));
+    }
+
+    /**
+     * Converts the expression to a BSON filter for a query.
+     * The format of the BSON is "{key: {$not: {$cmp: val}}}".
+     */
+    operator bsoncxx::document::view_or_value() const {
+        auto builder = bsoncxx::builder::basic::document{};
+        append_to_document_builder(builder);
+        return bsoncxx::document::view_or_value(builder);
```

Comment:
This is calling `builder`'s `operator bsoncxx::document::view()`. If builder gets cleaned up after we get a view, this might be an issue. I think you might need to do `builder.extract()`